### PR TITLE
connectors-ci: fix postgres integration testing

### LINKED ISSUE #25381: connectors-ci: source-postgres `integrationTestJava` with Dagger are failing
The `source-postgres` `integrationTestJava` are failing when run inside dagger with:

- `airbyte-ci connectors --name=source-postgres test`

38 tests are failing. A corresponding Gradle scan result can be found [here](https://scans.gradle.com/s/nqagenalmobjm). 
I think the error are due to connections problems between integration tests containers and `testcontainers` containers

Spotted errors:

`
{"type":"LOG","log":{"level":"ERROR","message":"ERROR i.a.w.i.DefaultAirbyteStreamFactory(internalLog):155 ERROR i.a.i.b.s.SshWrappedSource(read):76 Exception occurred while getting the delegate read iterator, closing SSH tunnel io.airbyte.commons.exceptions.ConnectionErrorException: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 10009ms.\n\tat io.airbyte.db.jdbc.DefaultJdbcDatabase.getMetaData(DefaultJdbcDatabase.java:87) ~[io.airbyte.airbyte-db-db-lib-23.0.1.jar:?]\n\tat io.airbyte.integrations.source.jdbc.AbstractJdbcSource.createDatabase(AbstractJdbcSource.java:425) ~[io.airbyte.airbyte-integrations.connectors-source-jdbc-23.0.1.jar:?]\n\tat io.airbyte.integrations.source.postgres.PostgresSource.createDatabase(PostgresSource.java:235) ~[io.airbyte.airbyte-integrations.connectors-source-postgres-23.0.1.jar:?]\n\tat io.airbyte.integrations.source.postgres.PostgresSource.createDatabase(PostgresSource.java:92) ~[io.airbyte.airbyte-integrations.connectors-source-postgres-23.0.1.jar:?]\n\tat io.airbyte.integrations.source.relationaldb.AbstractDbSource.read(AbstractDbSource.java:144) ~[io.airbyte.airbyte-integrations.connectors-source-relational-db-23.0.1.jar:?]\n\tat io.airbyte.integrations.base.ssh.SshWrappedSource.read(SshWrappedSource.java:74) ~[io.airbyte.airbyte-integrations.bases-base-java-23.0.1.jar:?]\n\tat io.airbyte.integrations.base.IntegrationRunner.runInternal(IntegrationRunner.java:141) ~[io.airbyte.airbyte-integrations.bases-base-java-23.0.1.jar:?]\n\tat io.airbyte.integrations.base.IntegrationRunner.run(IntegrationRunner.java:101) ~[io.airbyte.airbyte-integrations.bases-base-java-23.0.1.jar:?]\n\tat io.airbyte.integrations.source.postgres.PostgresSource.main(PostgresSource.java:485) ~[io.airbyte.airbyte-integrations.connectors-source-postgres-23.0.1.jar:?]\nCaused by: java.sql.SQLTransientConnectionException: HikariPool-1 - Connection is not available, request timed out after 10009ms.\n\tat com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:696) ~[HikariCP-5.0.1.jar:?]\n\tat com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:181) ~[HikariCP-5.0.1.jar:?]\n\tat com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:146) ~[HikariCP-5.0.1.jar:?]\n\tat com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100) ~[HikariCP-5.0.1.jar:?]\n\tat io.airbyte.db.jdbc.DefaultJdbcDatabase.getMetaData(DefaultJdbcDatabase.java:78) ~[io.airbyte.airbyte-db-db-lib-23.0.1.jar:?]\n\t... 8 more\nCaused by: org.postgresql.util.PSQLException: The connection attempt failed.\n\tat org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:331) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:247) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.Driver.makeConnection(Driver.java:434) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.Driver.connect(Driver.java:291) ~[postgresql-42.5.1.jar:42.5.1]\n\tat com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138) ~[HikariCP-5.0.1.jar:?]\n\tat com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:359) ~[HikariCP-5.0.1.jar:?]\n\tat com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201) ~[HikariCP-5.0.1.jar:?]\n\tat com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:470) ~[HikariCP-5.0.1.jar:?]\n\tat com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:733) ~[HikariCP-5.0.1.jar:?]\n\tat com.zaxxer.hikari.pool.HikariPool$PoolEntryCreator.call(HikariPool.java:712) ~[HikariCP-5.0.1.jar:?]\n\tat java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]\n\tat java.lang.Thread.run(Thread.java:833) ~[?:?]\nCaused by: java.net.UnknownHostException: dockerhost-source-postgres-gradle-integration-tests\n\tat sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:567) ~[?:?]\n\tat java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327) ~[?:?]\n\tat java.net.Socket.connect(Socket.java:633) ~[?:?]\n\tat org.postgresql.core.PGStream.createSocket(PGStream.java:241) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.core.PGStream.<init>(PGStream.java:98) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:109) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:235) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49) ~[postgresql-42.5.1.jar:42.5.1]\n\tat org.postgresql.jdbc.PgConnection.<init>(PgConnection… |  
`

### PR DESCRIPTION
**Pull request by alafanechere**
## What

Closes #25381 
This PR is an attempt to make source-postgres integration java test pass in the Dagger pipeline context, and still making them work for original /test run or locally.

Original problems:
* The container under test (source-postgres) could not reach the testcontainers databases that are spawned by the java tests
* Flakyness on CDC tests.

### Fixing networking issues
There were inconsistencies in the way test set up the source-postgres configuration to connect to the test database.
The Java test need to access to test databases to seed it with test data. In this context the java test should resolve the database host/port targetting the docker host and the exposed ports.
The source-postgres need to connect to test databases to READ. In this context, to resolve the database host/port the internal container ip/exposed ports must be used.

```mermaid
flowchart TD
    subgraph master ["Dagger engine"]
        itj["Gradle integrationTestJava"]
        subgraph dockerHost ["Docker host"]
            cut["source-postgres"]
            testcontainer["Test postgres database"]
            cut --"Resolve with docker host internal ip addresses and mapped ports"-->testcontainer
        end
        itj --"Resolve with docker host hostname and exposed ports for INSERT"--> testcontainer
    end
```

#### In practice:
To fill the source-postgres config we get the container internal ip address/exposed port by using `HostPortResolver` utility class: 
```java
    config = Jsons.jsonNode(ImmutableMap.builder()
        .put(JdbcUtils.HOST_KEY, HostPortResolver.resolveHost(container))
        .put(JdbcUtils.PORT_KEY, HostPortResolver.resolvePort(container))
        .put(JdbcUtils.DATABASE_KEY, container.getDatabaseName())
        .put(JdbcUtils.SCHEMAS_KEY, List.of(NAMESPACE))
        .put(JdbcUtils.USERNAME_KEY, container.getUsername())
        .put(JdbcUtils.PASSWORD_KEY, container.getPassword())
        .put("replication_method", replicationMethod)
        .put(JdbcUtils.SSL_KEY, false)
        .put("is_test", true)
        .build());
```

To seed the test database we use the host/exposed port by using the `.getHost()` / `.getFirstMappedPort()` on the container object:

```java
    try (final DSLContext dslContext = DSLContextFactory.create(
        config.get(JdbcUtils.USERNAME_KEY).asText(),
        config.get(JdbcUtils.PASSWORD_KEY).asText(),
        DatabaseDriver.POSTGRESQL.getDriverClassName(),
        String.format(DatabaseDriver.POSTGRESQL.getUrlFormatString(),
            container.getHost(),
            container.getFirstMappedPort(),
            config.get(JdbcUtils.DATABASE_KEY).asText()),
        SQLDialect.POSTGRES)) {
      final Database database = new Database(dslContext);

      database.query(ctx -> {
        ctx.execute("CREATE TABLE id_and_name(id INTEGER  primary key, name VARCHAR(200));");
        ctx.execute("INSERT INTO id_and_name (id, name) VALUES (1,'picard'),  (2, 'crusher'), (3, 'vash');");
        ctx.execute("CREATE TABLE starships(id INTEGER primary key, name VARCHAR(200));");
        ctx.execute("INSERT INTO starships (id, name) VALUES (1,'enterprise-d'),  (2, 'defiant'), (3, 'yamato');");
        ctx.execute("SELECT pg_create_logical_replication_slot('" + SLOT_NAME_BASE + "', 'pgoutput');");
        ctx.execute("CREATE PUBLICATION " + PUBLICATION + " FOR ALL TABLES;");
        return null;
      });
```

### Fixing CDC test flakyness

Increasing the value of `INITIAL_WAITING_SECONDS` to `30` seconds in the connector config made tests more stable. I observed the flakyness locally and in dagger pipelines.

### Improving Gradle performance in Dagger Pipeline
I introduced the use of the Gradle S3 cache in Dagger pipeline to benefit from cross build caching in the CI.
Even if caching is enabled by default in our main `gradle.properties` file the cache is not used for local run according to `settings.gradle`. To enable the use of the gradle remote cache I added the required env var to the gradle container. 

## 🚨 User Impact 🚨
According to my manual tests these changes make `source-postgres:integrationTestJava` pass locally, in dagger pipelines and we /test.
**Please note that these inconsistency in how we connect to test databases are also present on other java connectors. I focused on source-posgres because it's GA**. I might proceed in changing other connectors if this PR is 👍 and I spot other /test vs Dagger pipeline inconsistencies.

### CODE REVIEW

**Comment by alafanechere** [comment]
🎉Achieved both successful /test above and successful Dagger pipeline run [here](https://github.com/airbytehq/airbyte/actions/runs/4932858585/jobs/8816275912). Some CDC testing about WAL and Data type look flaky locally / on Dagger pipeline. I'm going to re-run the Dagger pipeline a couple of time for sanity.

**Comment by alafanechere** [comment]
Confirmed the flakyness of `CdcInitialSnapshotPostgresSourceDatatypeTest > testDataContent()` (failing in this [run](https://github.com/airbytehq/airbyte/actions/runs/4932953930/jobs/8816460013))

**Review by evantahler — COMMENTED** [review]
I'm going to punt on this review... I don't really know enough about what's going on in Java to be useful

File: tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py [hunk 1]
```diff
@@ -408,13 +408,24 @@ def with_gradle(
...
         .with_exec(["sh", "-c", "curl -fsSL https://get.docker.com | sh"])
         .with_exec(["mkdir", "/root/.gradle"])
-        .with_mounted_cache("/root/.gradle", root_gradle_cache, sharing=CacheSharingMode.LOCKED)
-        .with_exec(["rm", "-f", "/root/.gradle/caches/journal-1/journal-1.lock"])
         .with_exec(["mkdir", "/airbyte"])
         .with_mounted_directory("/airbyte", context.get_repo_dir(".", include=include))
         .with_mounted_cache("/airbyte/.gradle", airbyte_gradle_cache, sharing=CacheSharingMode.LOCKED)
         .with_workdir("/airbyte")
     )
+    if context.is_ci:
```
Comment (bnchrch) [inline tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py:416]: @alafanechere Can we leave a comment around this that explains
1. What we are doing
2. Why
3. When we should remove it and whos responsible

**Review by bnchrch — APPROVED** [review]
One comment about a comment.

Outside of that I have no objections.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/HostPortResolver.java [hunk 2]
```diff
@@ -12,14 +12,11 @@
 public class HostPortResolver {
 
   public static String resolveHost(GenericContainer container) {
-    return System.getProperty("os.name").toLowerCase().startsWith("mac")
-        ? getIpAddress(container)
-        : container.getHost();
+    return getIpAddress(container);
```
Comment (rodireich) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/HostPortResolver.java:15]: This is going to break the clickhouse integration tests on Mac I think.
The original change was #14701 back in July last year

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java [hunk 3]
```diff
@@ -60,11 +61,10 @@ protected void setupEnvironment(final TestDestinationEnv environment) throws Exc
     final JsonNode replicationMethod = Jsons.jsonNode(ImmutableMap.builder()
         .put("method", "Standard")
         .build());
-    final var containerOuterAddress = SshHelpers.getOuterContainerAddress(container);
-    final var containerInnerAddress = SshHelpers.getInnerContainerAddress(container);
+
     config = Jsons.jsonNode(ImmutableMap.builder()
-        .put("host", containerInnerAddress.left)
-        .put("port", containerInnerAddress.right)
+        .put("host", HostPortResolver.resolveHost(container))
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java:66]: Is this going to return the inner address and port of a container? This was the reason tests were originally failing when run on Mac.
Can you validate that running integration test on your Mac is still passing.
For example:
```
./gradlew --info :airbyte-integrations:connectors:source-postgres:integrationTestJava --tests "io.airbyte.integrations.io.airbyte.integration_tests.sources.PostgresSourceSSLCaCertificateAcceptanceTest"
```

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/CdcInitialSnapshotPostgresSourceDatatypeTest.java [hunk 4]
```diff
@@ -23,7 +23,7 @@ public class CdcInitialSnapshotPostgresSourceDatatypeTest extends AbstractPostgr
   private static final String SCHEMA_NAME = "test";
   private static final String SLOT_NAME_BASE = "debezium_slot";
   private static final String PUBLICATION = "publication";
-  private static final int INITIAL_WAITING_SECONDS = 5;
+  private static final int INITIAL_WAITING_SECONDS = 30;
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/CdcInitialSnapshotPostgresSourceDatatypeTest.java:26]: Is the difference on dagger expected?

Comment (alafanechere) [reply airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/CdcInitialSnapshotPostgresSourceDatatypeTest.java:26]: I faced flaky results on my local machine too and increasing the waiting time mitigate these problems. I found a PR from Sergio on another connector that did it for the same reasons.

File: airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java (hunk 3 above)
Comment (alafanechere) [reply airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java:66]: Yes it passed, this is why I made the change...
Screen Shot 2023-05-11 at 19 34 13

Comment (rodireich) [reply airbyte-integrations/connectors/source-postgres/src/test-integration/java/io/airbyte/integrations/io/airbyte/integration_tests/sources/AbstractPostgresSourceSSLCertificateAcceptanceTest.java:66]: thanks 
I'm just making sure as I don't know dagger 😅

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/HostPortResolver.java (hunk 2 above)
Comment (rodireich) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/util/HostPortResolver.java:15]: source-clickhouse that is

**Comment by rodireich** [comment]
@alafanechere I see source-clickhouse failing locally on my Mac with and without your change (I wonder if this is happening to all or just on my setup).
I'm ok with merging this change since it doesn't seem to break anything that wasn't already broken.

**Comment by alafanechere** [comment]
> @alafanechere I see source-clickhouse failing locally on my Mac with and without your change (I wonder if this is happening to all or just on my setup).
I'm ok with merging this change since it doesn't seem to break anything that wasn't already broken.

🙏

**Pull request merged**